tests/e2e/utils/utils.go

```diff
@@ -206,3 +209,161 @@ func RunSQLTest(ctx context.Context, frontendIngressIP string, isDistributed boo
 
 	return nil
 }
+
+// RunFlowTest runs the greptimedb flow tests on the frontend ingress IP.
+func RunFlowTest(ctx context.Context, frontendAddr string) error {
+	cfg := mysql.Config{
+		Net:                  "tcp",
+		Addr:                 frontendAddr,
+		AllowNativePasswords: true,
+	}
+
+	db, err := sql.Open("mysql", cfg.FormatDSN())
+	if err != nil {
+		return err
+	}
+
+	conn, err := db.Conn(ctx)
+	if err != nil {
+		return err
+	}
+
+	createInputTable := `CREATE TABLE numbers_input (
+	    number INT,
+	    ts TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+	    PRIMARY KEY(number),
+	    TIME INDEX(ts)
+	);`
+	_, err = conn.ExecContext(ctx, createInputTable)
+	if err != nil {
+		return err
+	}
+
+	createOutputTable := `CREATE TABLE sum_number_output (
+	    sum_number BIGINT,
+	    start_window TIMESTAMP TIME INDEX,
+	    end_window TIMESTAMP,
+	    update_at TIMESTAMP,
+	);`
+	_, err = conn.ExecContext(ctx, createOutputTable)
+	if err != nil {
+		return err
+	}
+
+	createFlow := fmt.Sprintf(`CREATE FLOW test_flow SINK TO sum_number_output AS
+		SELECT sum(number) FROM numbers_input GROUP BY tumble(ts, '5 second', %d);`, timestampInMillisecond())
+	_, err = conn.ExecContext(ctx, createFlow)
+	if err != nil {
+		return err
+	}
+
+	insertDataSQL := `INSERT INTO numbers_input(number, ts) VALUES (?, ?);`
+	selectDataSQL := `SELECT sum_number FROM sum_number_output;`
+
+	for i := 0; i < 20; i++ {
+		_, err = conn.ExecContext(ctx, insertDataSQL, i, timestampInMillisecond())
+		if err != nil {
+			return err
+		}
+	}
+
+	// Wait for the flow to process the data.
+	time.Sleep(5 * time.Second)
```

Comment:
Why sleep 5 second ?

Reply:
There is a small time window before the flownode receives the mirror requests and processes the data. Adding the sleeping logic will make sure we can always read the data.